Doc for layers and loading

TheBB opened the pull request:
I was encouraged to make a PR for this, so I rewrote it a little in a way that I feel it fits better.

Technically this file explains two things that are related but different
- how loading works (in Spacemacs as a whole, and Emacs)
- rules of thumb for how to write good layers

The structured writer in me would like to separate them more, but since we get so many questions about precisely these things I feel they should go together anyhow. It is of course not our job to explain Emacs internals, but I feel some overlap between our docs and the Emacs docs can go a long way.

TheBB commented:
How do I get that cute keyboard markup in tildes?

syl20bnr commented:
This is done here: https://github.com/syl20bnr/spacemacs/blob/develop/core/core-funcs.el#L112-L144

person808 commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,530 @@
+  (spacemacs|defvar-company-backends yoyo-mode)
+#+end_src
+
+This creates a variable called =company-backends-python-mode=. In the package
```
Should be `yoyo-mode` not `python-mode` to match

person808 commented:
`VIMUSERS.org` also has some sections on layers so you may want to take a look at it and add a link to `LAYERS.org`.

TheBB commented:
Thanks, updated. @syl20bnr I meant the Github rendered page, but I see now that none of the other pages have them. Was I hallucinating that?

syl20bnr commented:
It was the case back in the time when docs were in markdown ;-)

swaroopch commented:
+1, enlightening documentation.

TheBB commented:
I guess we should keep this open for a while. When the dust has settled from recent developments I'll update.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+contains /at least one/ of these files, ordered vaguely by importance.
+
+- =packages.el= :: for specifying which packages the layer uses
+- =extensions.el= :: for specifying extensions
```
You can remove all the mentions about extensions.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
+=<layer>/extensions/<extension>/= containing the source code for that extension.
```
The folder will be renamed to "local" instead of "extensions". The new name will become mandatory in 0.105.0 (in about 2 months).

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+** packages.el
+This file should define a variable called =<layer>-packages=, which should be a
+list of all the packages that this layer needs. Additionally, a layer may
+/exclude/ packages through the variable =<layer>-excluded-packages=. This will
```
`<layer>-excluded-packages` will be deprecated in 0.105.0, it is replaced by the property `:excluded` in the package list.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
```
`extensions` will be referenced as `local packages` starting 0.104.0.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+
+1. =<layer>/pre-init-<package>=
+2. =<layer>/init-<package>=
+3. =<layer>/post-init-<package>=
```
To avoid confusion between spacemacs `init` functions and `init` from `use-package` and `use-package` hooks (`:pre-init` and `:post-init`). All `init` function will be renamed to `configure`. Will become mandatory in 0.105.0 as well.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+in the load path for you.
+
+*Note:* A package will not be installed unless at least one layer defines an
+=init= function for it. That is to say, in a certain sense, the =init= function
```
Now only one layer can define the `configure` function for a package. This layer is called the owner of the package. In the case of multiple `configure` function, the last encountered one is declared the owner.

syl20bnr replied:
Still need to be fixed.

syl20bnr replied:
In fact you are right, the at least one is correct.
I guess we need a paragraph on the package ownership.

syl20bnr replied:
Scrap that........... I should read the whole article before commenting ;-D

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+setup. This can be used for managing cross-layer dependencies, which we will
+discuss later.
+
+** extensions.el
```
This section is not needed anymore. Should be replaced by a section on a local packages which correspond to packages with the property `:location local`.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+1. A handful packages are installed and loaded as part of the bootstrap process.
+   Among these packages, of particular interest to us, is =use-package=. All
+   subsequent code in the loading process may make use of it.
+2. Spacemacs goes through all the enabled layers and evaluates their files
```
goes through all used layers in order of declaration. So it is possible to use this property to override a `configure` function and steal the ownership of a package in a custom layer.

TheBB replied:
The `spacemacs` layer goes first, then?

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+layer does *not* own a package if it only defines =pre-init= or =post-init=
+functions.
+
+Best practices dictate that only /one/ layer should own any given package.
```
Now effectively only one layer can take the ownership of a package.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,540 @@
+Spacemacs installation by excluding the wrong packages, and it is not our
+intention to prevent this at all costs.
+
+*** Don't assume a fixed order
```
Actually packages are loaded in alphabetical order so it is always possible to know in advance if A will be configured before B.

TheBB replied:
But is that behaviour which is documented and guaranteed, or is it simply undefined behaviour that happens to be this way?

syl20bnr replied:
packages are explicitly sorted before being configured.

TheBB commented:
Wow, thanks a lot @syl20bnr. Saves me from having to figure it out. Hopefully I can get to this tonight.

TheBB commented:
Updated.

TheBB commented:
Github rendering can't deal with footnotes. Is that a deal breaker?

syl20bnr commented:
@TheBB I can live with it :-)

In the end I will keep `init` instead of `configure` for the ... `init` function. Sorry for the wrong indication.

My motivation about renaming the init functions (all of them, not only `init`, but also the `pre` and `post` variants) was to reduce the possible confusion between spacemacs init functions and use-package `:init` keyword. Turns out that `configure` is as confusing since it is like `:config`. Moreover according to use-package terminology, init functions are executed right away at initialization time so it would not make sense to rename the functions to `configure`. Last, it put too much burden on spacemacs users.

For these reasons I won't rename the init functions.

TheBB commented:
Okay :-P

Updated and rebased. (It said it couldn't be automatically merged... idk why, the rebase was automatic.)

TheBB commented:
Updated with `recipe`.

TheBB commented:
I just shat on the bed with this. I'll fix it...

TheBB commented:
Okay, we're back in business. Updated with the latest developments.

TheBB commented:
I added a note about the new layer dependency functionality.

syl20bnr commented on doc/DOCUMENTATION.org:
```diff
@@ -310,12 +313,13 @@ Where:
-repository, and =Extensions= are generally elisp code from git submodules.
+repository, and =Extensions= are generally elisp code from git submodules, or
+other code one might want to use but which isn't available on an =ELPA=
+compliant repository, for whatever reason.
```
This section is out of date. I think we can remove it. The difference between packages will be described somewhere else.

The file tree of a layer must be updated. There is no `extensions.el` anymore and `extensions` folder is now `local`.

Although these changes are out of scope of this PR I'm OK to make them here.

syl20bnr replied:
Or scratch that, I or another contributor will open a new PR to update all this stuff in DOCUMENTATION.org

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+
+- Package :: A set of Emacs Lisp files that, taken together, provide some
+     feature. Packages may be available on a package repository, such as ELPA or
+     MELPA.
```
as well as a service provider like github or even local.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+#+end_src
+
+This arranges for the relevant code to be executed after Helm is loaded (using
+eithe =require= or an autoload), or if Helm is already loaded, the code is
```
typo eithe -> either

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+
+** Use-package
+For /end users/ who are trying to put together an efficient Emacs configuration,
+there is a very useful /package/ called use-package that provides a macro called
```
you can emphasis use-package -> `=use=package=`

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+Additionally, for each local package (see the next section), there should be a
+folder =<layer>/local/<package>/= containing the source code for that package.
+Before initializing that package, Spacemacs will add this folder to the load
+path for you, so that you can treat it as any other package from MELPA.
```
I would remove `, so that you can treat it as any other package from MELPA`.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+     it.
+   Alternatively, if a package is part of the end user's
+   =dotspacemacs-additional-packages=, it will also be installed.
+4. All packages which should be installed are, and all packages which are
```
I think `used by the user` is missing after `are`.

TheBB replied:
I meant *All packages which should be installed are installed*, but I guess I can write that explicitly.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+For layers that require another layers to be enabled, use the functions
+=configuration-layer/declare-layer= and =configuration-layer/declare-layers= to
+ensure that layers are enabled even if the user has not enabled them explicitly.
+Calls to these functions should go in the =config.el= file.
```
should -> must :-)

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+in the =init-<package>= function, calls to =spacemacs|use-package-add-hook=
+typically happen in the =pre-init-<package>= functions, and not in
+=post-init-<package>=.
+
```
You can add that, as a rule of thumb, putting use-package hooks in `pre-init` is pretty safe so it should be the default choice when hooking stuff to use-package.

syl20bnr commented on doc/LAYERS.org:
```diff
@@ -0,0 +1,588 @@
+and in most cases you should not depend on the former, either. Preferably, write
+your layer so that it is independent of load ordering. The =pre=- and
+=post-init= functions are helpful, together with =configuration-layer/package-usedp=.
+
```
You are right about this since stuff can be lazy loaded but it is guaranteed that packages are configured alphabetically so when you know what you are doing you can rely on this property. It is the cases at several places in Spacemacs.

But maybe we should not document this to keep things simple. Relying on configuration order may be considered to be some kind of optimization we can do post merge of a PR.

syl20bnr commented:
I love this doc :+1:

TheBB commented:
Thanks, @syl20bnr. I have rebased and updated. I left the part about load order intact with a warning that it shouldn't be relied upon unless necessary.

Hopefully this is the last iteration of this PR. :-)

syl20bnr commented:
It's OK for me, nice work :+1:

syl20bnr commented:
This wonderful doc is now merged in develop.
Thank you! :heart:
